Add Splash screen

TheStonedTurtle opened the pull request:
Revised version of #3497. Related to runelite/launcher#64

![D20UsA5R0B](https://user-images.githubusercontent.com/29030969/61264294-dda78680-a740-11e9-96f3-327b75f0ab13.gif)


There's also a commit that adds client-sided launcher version check. This prevents the client from loading if launcher is outdated.

<details>
<p>

splash-screen:
![https://i.imgur.com/eqFdk2E.png](https://i.imgur.com/eqFdk2E.png)
no-splash:
![https://i.imgur.com/9nOR1AG.png](https://i.imgur.com/9nOR1AG.png)

</p>
</details>

The Splash Screen code can be found here:
<https://github.com/TheStonedTurtle/splashscreen>

deathbeam commented on runelite-client/src/main/java/net/runelite/client/RuneLite.java:
```diff
@@ -212,6 +220,32 @@ public ClientUpdateCheckMode convert(String v)
+			RuneLiteSplashScreen.setTheme();
+			if (options.has("no-splash"))
+			{
+				SPLASH_SCREEN = new RuneLiteSplashScreen(new File(LOGS_DIR, "client.log"));
```
this new File(LOGS_DIR, "client.log") should be just LOG_FILE constant at top

deathbeam replied:
also, honestly, these constructors do not make sense.Why is file not at first place in both scenarios, and why no-splash still needs splash screen, and why adding string to splash screen constructor changes behaviour of it?

TheStonedTurtle replied:
I create the splash screen even when no-splash is specified so I didn't have to null-check SPLASH_SCREEN at every `SPLASH_SCREEN.setMessage()`. It also lets me use the splash screens dialog popups which i think should be displayable regardless of if the splash screen is being displayed.

deathbeam replied:
I would rather do proxy method and null checks. If you want just the dialog popups, separate them from splash screen

deathbeam commented on runelite-client/src/main/java/net/runelite/client/RuneLite.java:
```diff
@@ -212,6 +220,32 @@ public ClientUpdateCheckMode convert(String v)
+		if (launcherVersion != null &&
+			!EXPECTED_LAUNCHER_VERSION.equalsIgnoreCase(launcherVersion))
+		{
+			SwingUtilities.invokeAndWait(SPLASH_SCREEN::invalidVersion);
```
Why is invalidVersion part of SPLASH_SCREEN? will it be used also in launcher? if no, why is it there

TheStonedTurtle replied:
I created the version check for the launcher but in order to force people to download the updated launcher we needed a version check in the client (at least temporarily)

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/clocks/ClockManager.java:
```diff
@@ -57,7 +57,12 @@
+
+	ClockManager()
+	{
+		SwingUtilities.invokeLater(() -> clockTabPanel = new ClockTabPanel(this));
```
This should be separate commit, and also explain why are you doing this

TheStonedTurtle replied:
The PR i linked had this change as it was causing a component creation off EDT issue for this file. It looks like this is no longer an issue though as I reverted this change and am not getting the error anymore. I don't understand why it's no longer an issue but I've reverted this change

TheStonedTurtle commented:
I've rebased the PR after #9428 was merged. I also updated the ClientUI error message referencing the log to use the splash screen error prompt.

![https://i.imgur.com/J3kDLso.png](https://i.imgur.com/J3kDLso.png)

abextm commented on runelite-client/src/main/java/net/runelite/client/RuneLite.java:
```diff
@@ -77,6 +80,14 @@
+	private static final RuneLiteProperties PROPERTIES = new RuneLiteProperties();
+	// Launcher was updated to check it's own version starting at v2.
+	// If they are using a version below 2 (Launcher 1.X.X) the client needs to force them to dl the updated launcher
+	private static final Pattern OUTDATED_LAUNCHER_VERSION = Pattern.compile("Launcher 1\\..*");
```
rather than having to do this with regex, which is difficult, it would probable be easier to just use strcmp with the last good version, and hope we don't need to release a 2 digit version number
